### PR TITLE
Add a skeleton kdcpolicy plugin

### PR DESCRIPTION
**Pull request by frozencemetery**
Signed-off-by: Robbie Harwood <rharwood@redhat.com>

Back in krb5-1.16 (and in RHEL-7.5), I added the [kdcpolicy plugin](http://web.mit.edu/kerberos/krb5-devel/doc/plugindev/kdcpolicy.html) to krb5.  This interface allows a module to hook all AS and TGS requests, potentially reject them, and manipulate ticket lifetimes.  This PR is a basic implementation of the interface, with all the plumbing IPA needs to get it loaded and installed.

There are ~~two~~ three use cases I had in mind, though of course many more are possible (this is a very powerful place to have a hook into the KDC):

- Reduced ticket lifetimes based on [auth indicator](http://web.mit.edu/kerberos/krb5-devel/doc/admin/auth_indicator.html)
- Adding (well, subtracting) random jitter from certain principal lifetimes to reduce contention from groups of tickets all needing renewal simultaneously
- Time-based access control to hosts/services

Since presumably we don't want any of that to be hardcoded behavior, the difficult part is now making it all configurable.  (As well as figuring out any behavior we want to control at the moment).  Per IRC conversation, I'm opening this PR so that we have something to look at while we discuss that.

BACKPORT NOTE: there's a spec file change in this PR.

### CODE REVIEW

**Comment by frozencemetery** [comment]
(I don't think the CI failures are caused by this change, but please correct me if that's wrong.)

**Comment by frozencemetery** [comment]
Fixed CI.  Thanks for pointing me at the failure.

**Comment by tiran** [comment]
You are using augeas to modify krb5 configuration. AFAIK we decided against augeas because augeas cannot handle all edge cases of krb5.conf.

**Comment by frozencemetery** [comment]
This is how the certauth plugin gets set up.  I'm happy to move that as well, if you like.  It makes sense to me to put this in a snippet in krb5.conf.d.

Problem is, there's no hook that I can see right now that upgrades *clients*.  Conceptually, these plugins are "server stuff" (though that lives in krb5.conf because not all plugins are server plugins).  While I can work around this for the server plugins, it means that the /etc/krb5.conf.d/freeipa file doesn't get updated (which means SPAKE doesn't get enabled on IPA client upgrade).

Suggestions?

**Comment by frozencemetery** [comment]
Answering my own question.  First commit makes it so that the snippet is re-written on update.  Second commit is the kdcpolicy plugin, now using the snippet.  Third commit moves the certauth plugin into the snippet by analogy. @tiran, ready for review (assuming it passes CI).

**Comment by frozencemetery** [comment]
test_forced_client_enrolment appears to be a DNF failure?  I guess you already know I have a lot of trouble figuring out what went wrong in the CI. I can't see any logs for test_advise or replica_promotion.

File: client/share/freeipa.template [hunk 1]
```diff
@@ -1,2 +1,12 @@
...
+
+[plugins]
+    kdcpolicy = {
+        module = ipakdb:kdb/ipadb.so
+        enable_only = ipakdb
+    }
+    certauth = {
+        module = ipakdb:kdb/ipadb.so
+        enable_only = ipakdb
+    }
```
Comment (rcritten) [inline client/share/freeipa.template]: Isn't ipadb.so only shipped in servers now?

File: daemons/ipa-kdb/ipa_kdb_kdcpolicy.c [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
```
Comment (rcritten) [inline daemons/ipa-kdb/ipa_kdb_kdcpolicy.c]: nit: the copyright header we use these days looks like:
```
/*
 * Copyright (C) <year>  FreeIPA Contributors see COPYING for license
 */```

File: daemons/ipa-kdb/ipa_kdb_kdcpolicy.c [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+#include <errno.h>
+#include <syslog.h>
+#include <krb5/kdcpolicy_plugin.h>
+
+#include "ipa_krb5.h"
+#include "ipa_kdb.h"
+
+struct krb5_kdcpolicy_moddata_st {
+    ;
+};
```
Comment (rcritten) [inline daemons/ipa-kdb/ipa_kdb_kdcpolicy.c]: An empty structure? Is this some krb5-ism?

File: freeipa.spec.in [hunk 4]
```diff
@@ -1241,10 +1241,12 @@ if [ $1 -gt 1 ] ; then
             cp /etc/ipa/ca.crt /var/lib/ipa-client/pki/kdc-ca-bundle.pem
             cp /etc/ipa/ca.crt /var/lib/ipa-client/pki/ca-bundle.pem
         fi
+
+        %{python} -c 'from ipaclient.install.client import configure_krb5_snippet; configure_krb5_snippet()' >>/var/log/ipaupgrade.log 2>&1
     fi
```
Comment (rcritten) [inline freeipa.spec.in]: We might want/need an ipa-client-upgrade script to do this instead, if at least for other distros (perhaps outside the scope of this PR).

**Comment by rcritten** [comment]
How/where would policy eventually be defined, or is that out-of-scope of this PR?

File: freeipa.spec.in (hunk 4 above)
Comment (frozencemetery) [reply freeipa.spec.in]: That was my thought as well.  We'd probably want to include the cert stuff above it if we ever make such a tool.

File: daemons/ipa-kdb/ipa_kdb_kdcpolicy.c (hunk 3 above)
Comment (frozencemetery) [reply daemons/ipa-kdb/ipa_kdb_kdcpolicy.c]: Ah, no.  It's places where we could store IPA state.  I'll add a comment.

**Comment by frozencemetery** [comment]
That's sort of the question I was hoping having code in hand would answer.

I assume configuration of this kind has to live in LDAP?  I don't know where though.  We also will need an interface to modify it.

**Comment by rcritten** [comment]
Right, it's sort of a chicken-and-egg problem I guess but there seems to be lack of context around this. It should probably be included as part of some initiative to actually include policy in the server.

This PR would add a bunch of configuration, running code, etc to a new install that effectively is a no-op (hopefully). I just don't know where/how to land this as-is.

File: client/share/freeipa.template (hunk 1 above)
Comment (tiran) [reply client/share/freeipa.template]: Yes, ``ipadb.so`` is only available in ``ipa-server`` package. Please move the snippet to a server specific template.

**Comment by frozencemetery** [comment]
Updated and rebased.  (Note that it's not technically an issue to have such configuration on the client, but you're right it's cleaner this way.)

**Comment by tiran** [comment]
Installation is failing with

```
2019-04-11T22:37:29Z DEBUG Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/ipaserver/install/service.py", line 603, in start_creation
    run_step(full_msg, method)
  File "/usr/lib/python3.7/site-packages/ipaserver/install/service.py", line 589, in run_step
    method()
  File "/usr/lib/python3.7/site-packages/ipaserver/install/krbinstance.py", line 351, in __configure_instance
    self.__template_file(paths.KRB5_FREEIPA_SERVER)
  File "/usr/lib/python3.7/site-packages/ipaserver/install/krbinstance.py", line 321, in __template_file
    conf = ipautil.template_file(template, self.sub_dict)
  File "/usr/lib/python3.7/site-packages/ipapython/ipautil.py", line 318, in template_file
    with open(infilename) as f:
FileNotFoundError: [Errno 2] No such file or directory: '/usr/share/ipa/freeipa-server.template'
```

File: install/share/freeipa-server.template [hunk 5]
```diff
@@ -0,0 +1,9 @@
+[plugins]
```
Comment (tiran) [inline install/share/freeipa-server.template]: The build system doesn't pick up new files automatically. You have to add this file to ``Makefile.am`` and maybe ``freeipa.spec.in``

**Comment by frozencemetery** [comment]
Closing this to avoid confusion since it's been subsumed by #3358.

**Pull request closed without merging**